File: study/lua/src/app/fight/report/StatisticsTotal.lua
local BaseData = require("app.data.BaseData")
local StatisticsTotal = class("StatisticsTotal", BaseData)

local schema = {}
schema["unitStatistics"] = {"table", {}}
schema["petStatistics"] = {"table", {}}
StatisticsTotal.schema = schema
local HeroSkillEffect = require("app.config.hero_skill_effect")

function StatisticsTotal:ctor()
    StatisticsTotal.super.ctor(self)
    self._buffTable = {} --记录一下globelID，用于计算伤害
    self._buffTalbePet = {} --神兽buff的globelID
end

function StatisticsTotal:clear()
    self:setUnitStatistics({})
    self:setPetStatistics({})
end

function StatisticsTotal:parseStatistics(waveData, reportData)
    self._leftName = reportData:getLeftName()
    self._rightName = reportData:getRightName()
    self._leftOfficerLevel = reportData:getAttackOfficerLevel()
    self._rightOfficerLevel = reportData:getDefenseOfficerLevel()
    local unitList = waveData:getUnits()
    local finalUnit = waveData:getFinalUnits()
    self:initStatistics(unitList, finalUnit)

    local pets = waveData:getPets()
    self:initPetsStatistics(pets)

    local roundList = waveData:getRounds()
    for i = 1, #roundList do
        local roundData = roundList[i]
        for i = 1, #roundData.attacks do
            local attackData = roundData.attacks[i]
            self:_parseOneAttack(attackData)
        end
    end
end

function StatisticsTotal:_addAngerByConfig(configId, count, statistics)
    local configData = HeroSkillEffect.get(configId)
    assert(configData, "not buff effect id = " .. configId)
    if configData.buff_sum ~= 0 then
        statistics:addStatistics(configData.buff_sum, count, configData.buff_sum_txt)
    end
end

function StatisticsTotal:_calcAnger(attack)
    --之后要根据angerdata里面的id来选择统计的人头
    local attackId = attack.stageId
    local statistics = self:getUnitStatisticsById(attackId)
    for i = 1, #attack.angers do
        local data = attack.angers[i]
        if data.showType ~= 0 then
            self:_addAngerByConfig(data.configId, data.value, statistics)
        end
    end
end

function StatisticsTotal:_calcDamage(attack)
    local statistics = self:getUnitStatisticsById(attack.stageId)
    for i = 1, #attack.targets do
        local attackInfo = attack.targets[i]
        local damage = attackInfo.damage
        statistics:updateValue(damage.type, damage.value)
    end
    for i = 1, #attack.addTargets do
        local addAttackInfo = attack.addTargets[i]
        local damage = addAttackInfo.damage
        statistics:updateValue(damage.type, damage.value)
    end
end

function StatisticsTotal:_calcBattleEffects(attack)
    for i = 1, #attack.battleEffects do
        local data = attack.battleEffects[i]
        local statistics = self:getUnitStatisticsById(data.attackId)
        local damage = data.damage
        statistics:updateValue(damage.type, damage.value)
    end
end

--添加buff统计
function StatisticsTotal:_calcBuffs(attack)
    for i = 1, #attack.addBuffs do
        local data = attack.addBuffs[i]
        local atkId = data.attackId
        if atkId and atkId == 0 then
            atkId = attack.stageId
        end
        local statistics = self:getUnitStatisticsById(atkId)
        assert(
            statistics,
            "error attack statistics info " .. tostring(attack.isPet) .. attack.petCamp .. "   " .. atkId
        )
        self._buffTable[data.globalId] = atkId
        local configData = HeroSkillEffect.get(data.configId)
        assert(configData, "not buff effect id = " .. data.configId)
        if configData.buff_sum ~= 0 then
            statistics:addStatistics(configData.buff_sum, 1, configData.buff_sum_txt)
        end
    end
end

--buff伤害统计
function StatisticsTotal:_calcBuffEffect(attack)
    for i = 1, #attack.buffEffects do
        local data = attack.buffEffects[i]
        local attackUnit, attackPet, configId = self:_getStatisticsUnitByBuffGlobalId(data.globalId)
        if attackUnit then
            -- assert(attackUnit, "no unit has global id = "..data.globalId)
            local damage = data.damage
            attackUnit:updateValue(damage.type, damage.value)
        end
        if attackPet then
            local configData = HeroSkillEffect.get(configId)
            assert(configData, "not buff effect id = " .. configId)
            if configData.buff_sum ~= 0 then
                attackPet:addStatistics(configData.buff_sum, 1, configData.buff_sum_txt)
            end
        end
    end
end

function StatisticsTotal:_parsePetAttack(attack)
    self:_calcPetAnger(attack)
    self:_calcPetBuffs(attack)
    self:_calcBuffEffect(attack)
    self:_calcBattleEffectsPet(attack)
end

function StatisticsTotal:_calcBattleEffectsPet(attack)
    for i = 1, #attack.battleEffects do
        local data = attack.battleEffects[i]
        local petCamp = math.floor(data.attackId / 100)
        local petId = data.petId
        local petStatistics = self:getPetStatisticsbyId(petCamp, petId)
        local configData = HeroSkillEffect.get(data.configId)
        if configData.buff_sum ~= 0 then
            petStatistics:addStatistics(configData.buff_sum, 1, configData.buff_sum_txt)
        end
    end
end

function StatisticsTotal:_calcPetAnger(attack)
    for i = 1, #attack.angers do
        local camp = attack.petCamp
        local id = attack.petId
        local petStatistics = self:getPetStatisticsbyId(camp, id)
        local data = attack.angers[i]
        local tarCamp = math.floor(data.stageId / 100) --怒气受方的阵营
        if data.type == 2 and camp ~= tarCamp then --如果是加怒，并且加到对方阵营，那么不统计
            
        else
            if data.showType ~= 0 then
                local configData = HeroSkillEffect.get(data.configId)
                if configData.buff_sum ~= 0 then
                    petStatistics:addStatistics(configData.buff_sum, data.value, configData.buff_sum_txt)
                end
            end
        end
    end
end

function StatisticsTotal:_calcPetBuffs(attack)
    for i = 1, #attack.addBuffs do
        local data = attack.addBuffs[i]
        local camp = math.floor(data.attackId / 100)
        local id = data.petId
        local petStatistics = self:getPetStatisticsbyId(camp, id)
        if not petStatistics then 
            return 
        end
        self._buffTalbePet[data.globalId] = {
            camp,
            id,
            data.configId
        }
        local configData = HeroSkillEffect.get(data.configId)
        if configData and configData.buff_sum ~= 0 and configData.buff_sum ~= 8 then       --麒麟这边上buff的时候不统计
            petStatistics:addStatistics(configData.buff_sum, 1, configData.buff_sum_txt)
        end
    end
end

function StatisticsTotal:_parseOneAttack(attack)
    --神兽攻击不计入统计
    if attack.isPet then
        self:_parsePetAttack(attack)
        return
    end

    if attack.isHistory then
        return
    end

    --攻击目标的加减血
    self:_calcDamage(attack)
    --怒气统计
    self:_calcAnger(attack)
    --全局展示buff统计
    self:_calcBattleEffects(attack)
    --添加buff统计
    self:_calcBuffs(attack)
    --buff伤害统计
    self:_calcBuffEffect(attack)
end

function StatisticsTotal:getUnitStatisticsById(stageId)
    local list = self:getUnitStatistics()
    for i, v in pairs(list) do
        if stageId == v:getStageId() then
            return v
        end
    end
end

function StatisticsTotal:getPetStatisticsbyId(camp, petId)
    local list = self:getPetStatistics()
    for i, v in pairs(list) do
        if camp == v:getPetCamp() and petId == v:getConfigData().id then
            return v
        end
    end
end

function StatisticsTotal:initStatistics(unitList, finalUnits)
    local list = {}
    for _, v in pairs(unitList) do
        local name = self._leftName
        local officerLevel = self._leftOfficerLevel
        if math.floor(v.stageId / 100) == 2 then
            name = self._rightName
            officerLevel = self._rightOfficerLevel
        end
        local isAlive = false
        for _, unit in pairs(finalUnits) do
            if unit.stageId == v.stageId then
                isAlive = true
                break
            end
        end
        local newStatistics =
            require("app.fight.report.UnitStatisticsData").new(
            v.stageId,
            v.configId,
            name,
            v.monsterId,
            v.rankLevel,
            officerLevel,
            isAlive,
            v.isLeader,
            v.limitLevel,
            v.limitRedLevel
        )
        table.insert(list, newStatistics)
    end
    self:setUnitStatistics(list)
end

function StatisticsTotal:initPetsStatistics(petList)
    local list = {}
    for _, v in pairs(petList) do
        local pet = require("app.fight.report.UnitPetStatisticsData").new(v.camp, v.configId, v.star)
        table.insert(list, pet)
    end
    self:setPetStatistics(list)
end

function StatisticsTotal:_getStatisticsUnitByBuffGlobalId(globalId)
    -- local id = self._buffTable[globalId]
    -- local camp, petId = self._buffTalbePet[globalId]

    local isPet = false
    local buffConfig = nil
    local tbl = self._buffTable[globalId]
    if not tbl then
        tbl = self._buffTalbePet[globalId]
        isPet = true
    end
    if not tbl then 
        return 
    end
    local unit, unitPet
    if isPet then
        unitPet = self:getPetStatisticsbyId(tbl[1], tbl[2])
        buffConfig = tbl[3]
    else
        unit = self:getUnitStatisticsById(id)
    end
    return unit, unitPet, buffConfig
end

function StatisticsTotal:getMaxDamage()
    local maxDamage = 0
    local list = self:getUnitStatistics()
    for i, v in pairs(list) do
        local damage = v:getStatisticsDamage()
        if damage > maxDamage then
            maxDamage = damage
        end
    end
    return maxDamage
end

-- function StatisticsTotal:getMaxDamage(camp)
--     local maxDamage = 0
--     local list = self:getUnitStatistics()
--     for i, v in pairs(list) do
--         if camp == math.floor(v:getStageId()/100) then
--             -- local damage = v:getDamage()
--             -- if v:getConfigData().sum_type == 1 then
--             --     damage = v:getHeal()
--             -- end
--             local damage = v:getStatisticsDamage()
--             if damage > maxDamage then
--                 maxDamage = damage
--             end
--         end
--     end
--     return maxDamage
-- end

function StatisticsTotal:getDataListByCamp(camp)
    local ret = {}
    local list = self:getUnitStatistics()
    for i, v in pairs(list) do
        if math.floor(v:getStageId() / 100) == camp then
            table.insert(ret, v)
        end
    end
    table.sort(
        ret,
        function(a, b)
            return a:getStageId() < b:getStageId()
        end
    )
    return ret
end

function StatisticsTotal:getPetDataListByCamp(camp)
    local ret = {}
    local list = self:getPetStatistics()
    for i, v in pairs(list) do
        if v:getPetCamp() == camp then
            table.insert(ret, v)
        end
    end
    table.sort(
        ret,
        function(a, b)
            return a:getPetId() > b:getPetId()
        end
    )
    return ret
end

return StatisticsTotal

File: study/ts/script/fight/report/StatisticsTotal.ts
import { Unit, Pet, Round, Attack } from "./WaveData";
import { UnitStatisticsData } from "./UnitStatisticsData";
import { UnitPetStatisticsData } from "./UnitPetStatisticsData";
import { G_ConfigLoader } from "../../init";
import { ConfigNameConst } from "../../const/ConfigNameConst";

export class StatisticsTotal {
    unitStatistics: UnitStatisticsData[] = [];
    petStatistics: UnitPetStatisticsData[] = [];

    _buffTable = [];
    _buffTalbePet = [];
    _leftName = "";
    _rightName = "";
    _leftOfficerLevel = 0;
    _rightOfficerLevel = 0;

    constructor() {
        this._buffTable = [] //记录一下globelID，用于计算伤害
        this._buffTalbePet = [] //神兽buff的globelID
    }


    public parseStatistics(units: Unit[], finalUnits: Unit[], pets: Pet[], rounds: Round[],
        leftName, rightName, attackOfficerLevel, defenseOfficerLevel) {
        this._leftName = leftName;
        this._rightName = rightName;
        this._leftOfficerLevel = attackOfficerLevel;
        this._rightOfficerLevel = defenseOfficerLevel;
        this.initStatistics(units, finalUnits);
        this.initPetsStatistics(pets);

        for (let i = 0; i < rounds.length; i++) {
            let roundData = rounds[i];
            for (let j = 0; j < roundData.attacks.length; j++) {
                let attackData = roundData.attacks[j];
                this.parseOneAttack(attackData);
            }
        }
    }

    private addAngerByConfig(configId, count, statistics) {
        let configData = G_ConfigLoader.getConfig(ConfigNameConst.HERO_SKILL_EFFECT).get(configId);
        if (configData.buff_sum != 0)
            statistics.addStatistics(configData.buff_sum, count, configData.buff_sum_txt)
    }

    private calcAnger(attack: Attack) {// 之后要根据angerdata里面的id来选择统计的人头
        let attackId = attack.stageId
        let statistics = this.getUnitStatisticsById(attackId)
        for (let i = 0; i < attack.angers.length; i++) {
            let data = attack.angers[i];
            if (data.showType != 0) {
                this.addAngerByConfig(data.configId, data.value, statistics)
            }
        }
    }

    private calcDamage(attack: Attack) {
        // console.log("calcDamage:",attack.stageId);
        let statistics: UnitStatisticsData = this.getUnitStatisticsById(attack.stageId)

        for (let i = 0; i < attack.targets.length; i++) {
            let attackInfo = attack.targets[i];
            let damage = attackInfo.damage;
            statistics.updateValue(damage.type, damage.value);
        }

        for (let i = 0; i < attack.addTargets.length; i++) {
            let attackInfo = attack.addTargets[i];
            let damage = attackInfo.damage;
            statistics.updateValue(damage.type, damage.value);
        }
    }
    private calcBattleEffects(attack) {
        for (let i = 0; i < attack.battleEffects.length; i++) {
            let data = attack.battleEffects[i];
            let statistics = this.getUnitStatisticsById(data.attackId)
            let damage = data.damage;
            statistics.updateValue(damage.type, damage.value)
        }
    }

    // 添加buff统计
    private calcBuffs(attack) {
        for (let i = 0; i < attack.addBuffs.length; i++) {
            let data = attack.addBuffs[i];
            let atkId = data.attackId;
            if (atkId && atkId == 0)
                atkId = attack.stageId;
            let statistics = this.getUnitStatisticsById(atkId)
            this._buffTable[data.globalId] = atkId;
            let configData = G_ConfigLoader.getConfig(ConfigNameConst.HERO_SKILL_EFFECT).get(data.configId);
            if (configData.buff_sum != 0)
                statistics.addStatistics(configData.buff_sum, 1, configData.buff_sum_txt)
        }
    }

    //buff伤害统计
    private calcBuffEffect(attack) {
        for (let i = 0; i < attack.buffEffects.length; i++) {
            let data = attack.buffEffects[i];
            let unitData = this.getStatisticsUnitByBuffGlobalId(data.globalId);
            let attackUnit = unitData.unit;
            let attackPet = unitData.unitPet;
            let configId = unitData.buffConfig;
            if (attackUnit) {
                let damage = data.damage
                attackUnit.updateValue(damage.type, damage.value)
            }
            if (attackPet) {
                let configData = G_ConfigLoader.getConfig(ConfigNameConst.HERO_SKILL_EFFECT).get(configId);
                if (configData.buff_sum != 0)
                    attackPet.addStatistics(configData.buff_sum, 1, configData.buff_sum_txt)
            }
        }
    }

    private parsePetAttack(attack) {
        this.calcPetAnger(attack)
        this.calcPetBuffs(attack)
        this.calcBuffEffect(attack)
        this.calcBattleEffectsPet(attack)
    }

    private calcBattleEffectsPet(attack) {
        for (let i = 0; i < attack.battleEffects.length; i++) {
            let data = attack.battleEffects[i]
            let petCamp = Math.floor(data.attackId / 100)
            let petId = data.petId
            let petStatistics = this.getPetStatisticsbyId(petCamp, petId)
            let configData = G_ConfigLoader.getConfig(ConfigNameConst.HERO_SKILL_EFFECT).get(data.configId);
            if (configData.buff_sum != 0)
                petStatistics.addStatistics(configData.buff_sum, 1, configData.buff_sum_txt)

        }
    }
    private calcPetAnger(attack) {

        for (let i = 0; i < attack.angers.length; i++) {
            let camp = attack.petCamp;
            let id = attack.petId;
            let petStatistics = this.getPetStatisticsbyId(camp, id)
            let data = attack.angers[i]
            var tarCamp = Math.floor(data.stageId / 100);
            if (data.type == 2 && camp != tarCamp) {
            } else {
                if (data.showType != 0) {
                    let configData = G_ConfigLoader.getConfig(ConfigNameConst.HERO_SKILL_EFFECT).get(data.configId);
                    if (configData.buff_sum != 0)
                        petStatistics.addStatistics(configData.buff_sum, data.value, configData.buff_sum_txt)
                }
            }
        }
    }


    private calcPetBuffs(attack) {
        for (let i = 0; i < attack.addBuffs.length; i++) {
            let data = attack.addBuffs[i]
            let camp = Math.floor(data.attackId / 100)
            let id = data.petId
            let petStatistics = this.getPetStatisticsbyId(camp, id)
            if (!petStatistics) {
                return;
            }
            this._buffTalbePet[data.globalId] = [
                camp,
                id,
                data.configId
            ]
            let configData = G_ConfigLoader.getConfig(ConfigNameConst.HERO_SKILL_EFFECT).get(data.configId);
            if (configData && configData.buff_sum != 0 && configData.buff_sum != 8) //麒麟这边上buff的时候不统计
                petStatistics.addStatistics(configData.buff_sum, 1, configData.buff_sum_txt)

        }
    }

    private parseOneAttack(attack: Attack) {// 神兽攻击不计入统计
        if (attack.isPet) {
            this.parsePetAttack(attack)
            return
        }
        if (attack.isHistory)
            return
        // 攻击目标的加减血
        this.calcDamage(attack)
        // 怒气统计
        this.calcAnger(attack)
        // 全局展示buff统计
        this.calcBattleEffects(attack)
        // 添加buff统计
        this.calcBuffs(attack)
        // buff伤害统计
        this.calcBuffEffect(attack)
    }
    private getUnitStatisticsById(stageId) {
        let list = this.unitStatistics;
        for (let i = 0; i < list.length; i++) {
            if (stageId == list[i].stageId)
                return list[i];
        }
    }

    private getPetStatisticsbyId(camp, petId) {
        let list = this.petStatistics;
        for (let i = 0; i < list.length; i++) {
            if (camp == list[i].petCamp && petId == list[i].configData.id)
                return list[i];
        }
    }

    private initStatistics(unitList, finalUnits) {
        let list = [];

        for (let i = 0; i < unitList.length; i++) {
            let name = this._leftName
            let officerLevel = this._leftOfficerLevel
            if (Math.floor(unitList[i].stageId / 100) == 2) {
                name = this._rightName
                officerLevel = this._rightOfficerLevel
            }
            let isAlive = false

            for (let j = 0; j < finalUnits.length; j++) {
                if (finalUnits[j].stageId == unitList[i].stageId) {
                    isAlive = true;
                    break;
                }
            }
            let newStatistics = new UnitStatisticsData(
                unitList[i].stageId,
                unitList[i].configId,
                name,
                unitList[i].monsterId,
                unitList[i].rankLevel,
                officerLevel,
                isAlive,
                unitList[i].isLeader,
                unitList[i].limitLevel,
                unitList[i].limitRedLevel
            )
            list.push(newStatistics);
        }
        this.unitStatistics = list;
    }
    private initPetsStatistics(petList) {
        let list = [];
        for (let i = 0; i < petList.length; i++) {
            let pet = new UnitPetStatisticsData(petList[i].camp, petList[i].configId, petList[i].star)
            list.push(pet);
        }
        this.petStatistics = list;
    }

    private getStatisticsUnitByBuffGlobalId(globalId) {
        let id = this._buffTable[globalId]
        // local camp, petId = self._buffTalbePet[globalId]

        let isPet = false;
        let buffConfig = null;
        let tbl = this._buffTable[globalId]
        if (!tbl) {
            tbl = this._buffTalbePet[globalId]
            isPet = true
        }
        if (!tbl) {
            return;
        }
        let unit, unitPet;
        if (isPet) {
            unitPet = this.getPetStatisticsbyId(tbl[1], tbl[2])
            buffConfig = tbl[3];
        }
        else
            unit = this.getUnitStatisticsById(id);
        return { unit: unit, unitPet: unitPet, buffConfig: buffConfig }
    }
    public getMaxDamage() {
        let maxDamage = 0
        let list = this.unitStatistics;
        for (let i = 0; i < list.length; i++) {
            let damage = list[i].getStatisticsDamage();
            if (damage > maxDamage)
                maxDamage = damage
        }
        return maxDamage
    }

    public getDataListByCamp(camp) {
        let ret = []
        let list = this.unitStatistics;
        for (let i = 0; i < list.length; i++) {
            if (Math.floor(list[i].stageId / 100) == camp)
                ret.push(list[i]);

        }

        ret.sort(function (a, b) {
            return a.getStageId - b.getStageId;
        })

        return ret
    }

    public getPetDataListByCamp(camp) {
        let ret = []
        let list = this.petStatistics;

        for (let i = 0; i < list.length; i++) {
            if (list[i].petCamp == camp)
                ret.push(list[i]);
        }
        ret.sort(
            function (a, b) {
                return b.getPetId() - a.getPetId();
            });
        return ret;
    }
}